core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -245,63 +302,293 @@ public static BytesReference serializeDiffClusterState(Diff diff, Version nodeVe
+            super(in);
+        }
+
+        public FailedToCommitException(String msg, Object... args) {
+            super(msg, args);
+        }
+    }
+
+    class SendingController {
+
+        private final ClusterState clusterState;
+
+        public BlockingClusterStatePublishResponseHandler getPublishResponseHandler() {
+            return publishResponseHandler;
+        }
+
+        private final BlockingClusterStatePublishResponseHandler publishResponseHandler;
+        final ArrayList<DiscoveryNode> sendAckedBeforeCommit = new ArrayList<>();
+        final CountDownLatch committedOrFailedLatch;
+
+        // writes and reads of these are protected under synchronization
+        boolean committedOrFailed; // true if a decision was made w.r.t committing or failing
+        boolean committed;  // true if cluster state was committed
+        int neededMastersToCommit; // number of master nodes acks still needed before committing
+        int pendingMasterNodes; // how many master node still need to respond
+
+        // an external marker to note that the publishing process is timed out. This is useful for proper logging.
+        final AtomicBoolean publishingTimedOut = new AtomicBoolean();
+
+        private SendingController(ClusterState clusterState, int minMasterNodes, int totalMasterNodes, BlockingClusterStatePublishResponseHandler publishResponseHandler) {
+            this.clusterState = clusterState;
+            this.publishResponseHandler = publishResponseHandler;
+            this.neededMastersToCommit = Math.max(0, minMasterNodes - 1); // we are one of the master nodes
+            this.pendingMasterNodes = totalMasterNodes - 1;
+            if (this.neededMastersToCommit > this.pendingMasterNodes) {
+                throw new FailedToCommitException("not enough masters to ack sent cluster state. [{}] needed , have [{}]", neededMastersToCommit, pendingMasterNodes);
             }
+            this.committed = neededMastersToCommit == 0;
+            this.committedOrFailed = committed;
+            this.committedOrFailedLatch = new CountDownLatch(committed ? 0 : 1);
+        }
 
+        public void waitForCommit(TimeValue commitTimeout) {
+            boolean timedout = false;
             try {
-                listener.onNewClusterState(lastSeenClusterState, new NewClusterStateListener.NewStateProcessed() {
-                    @Override
-                    public void onNewClusterStateProcessed() {
-                        try {
-                            channel.sendResponse(TransportResponse.Empty.INSTANCE);
-                        } catch (Throwable e) {
-                            logger.debug("failed to send response on cluster state processed", e);
-                        }
-                    }
+                timedout = committedOrFailedLatch.await(commitTimeout.millis(), TimeUnit.MILLISECONDS) == false;
+            } catch (InterruptedException e) {
 
```

Comment:
Maybe we should restore interrupted state or at least explain why we ignore it here?